File: PostRollApp/Sources/Services/LayoutSidecar.swift
import Foundation

/// Where a rendered preview's layout sidecar lives (#267).
///
/// The collage and the Thursday reel strip each write a JSON file beside their
/// PNG recording the cell rectangles, so the crop editor can draw its controls
/// over the right parts of the image and the export compositor can honour a
/// dragged divider.
///
/// Python writes it and this side reads it, from three screens and the export
/// path. The name used to be rebuilt independently in five places, one of them
/// a hardcoded `reel_preview_layout.json` literal, with nothing forcing them to
/// agree. The first one changed would have broken the editor or the export on
/// one screen only while the others kept working, which is the hardest version
/// of this to notice.
///
/// `postroll/media/layout_sidecar.py` is the writing half.
/// `tests/fixtures/layout_sidecar.json` is the contract both satisfy.
enum LayoutSidecar {

    /// Appended to the preview's name, without its extension.
    static let suffix = "_layout.json"

    /// The sidecar that belongs to `preview`, in the same directory.
    static func url(for preview: URL) -> URL {
        preview
            .deletingLastPathComponent()
            .appendingPathComponent(preview.deletingPathExtension().lastPathComponent + suffix)
    }

    /// What a sidecar holds: the cells, and which collage design made them.
    ///
    /// `version` is nil for a sidecar written before the stamp existed (#160),
    /// which is a bare array. Nil rather than 0: not knowing which design made
    /// something is a different fact from knowing it was the first one.
    struct Contents: Equatable {
        var version: Int?
        var cells: [CollageCell]

        /// Where the branded centre strip SAT in this layout (#970).
        ///
        /// Nil for a reel strip, which has no strip, and for every collage
        /// sidecar written before this was recorded. Nil means "not recorded"
        /// rather than "no strip": the checks that read it decline to judge the
        /// band rather than inventing one, because a made up band would refuse
        /// layouts for a position nobody chose.
        var strip: StripBand?

        /// Whether this collage predates the design this build renders.
        var isStale: Bool { (version ?? 0) < CollageDesign.collageDesignVersion }
    }

    struct StripBand: Equatable, Decodable {
        var y: Int
        var h: Int
    }

    /// Read a sidecar in either shape, tolerating everything already on disk.
    ///
    /// Every collage rendered before the stamp existed has the old bare-array
    /// shape, and treating those as unreadable would blank a day's crop
    /// editing rather than badge it.
    ///
    /// Never throws: a missing or corrupt sidecar means the editor falls back
    /// to the automatic layout, which is what it did before any of this.
    static func read(at url: URL) -> Contents {
        guard let data = try? Data(contentsOf: url) else { return Contents(version: nil, cells: []) }
        if let cells = try? JSONDecoder().decode([CollageCell].self, from: data) {
            return Contents(version: nil, cells: cells)
        }
        struct Envelope: Decodable {
            var version: Int?
            var cells: [CollageCell]?
            var strip: StripBand?
        }
        guard let envelope = try? JSONDecoder().decode(Envelope.self, from: data) else {
            return Contents(version: nil, cells: [])
        }
        // A band of zero height is not a band. Python omits `strip` entirely
        // where there is none, so a zero here is a malformed file rather than a
        // layout without a strip, and reading it as one would judge every cell
        // against a band with no thickness (L257).
        let band = envelope.strip.flatMap { $0.h > 0 ? $0 : nil }
        return Contents(version: envelope.version, cells: envelope.cells ?? [],
                        strip: band)
    }

    /// The contents of the sidecar belonging to a rendered preview.
    static func read(forPreview preview: URL) -> Contents {
        read(at: url(for: preview))
    }
}

File: postroll/media/layout_sidecar.py
"""Where a rendered preview's layout sidecar lives.

The collage and the Thursday reel strip each write a JSON file beside their
PNG recording the cell rectangles, so the app can draw crop controls over the
right parts of the image. The name is derived from the PNG's own name, and that
derivation is here rather than in each writer, because Swift reads the same
file and rebuilt the name in five separate places (#267).

`PostRollApp/Sources/Services/LayoutSidecar.swift` is the reading half.
`tests/fixtures/layout_sidecar.json` is the contract both satisfy.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from .design_tokens import COLLAGE_DESIGN_VERSION


#: Appended to the preview's name, without its extension.
SUFFIX = "_layout.json"


def layout_sidecar_path(preview: Path | str) -> Path:
    """The sidecar that belongs to `preview`, in the same directory."""
    preview = Path(preview)
    return preview.parent / (preview.stem + SUFFIX)


def write_layout_sidecar(path: Path | str, cells: list[dict[str, Any]],
                         version: int = COLLAGE_DESIGN_VERSION,
                         strip: tuple[int, int] | None = None) -> None:
    """Write the cells plus the design version that produced them (#160).

    The file used to be a bare array. The envelope is what lets the app tell a
    collage rendered by the current design from one rendered by an older one,
    which it previously could not, so a redesign left every existing collage
    showing the old look until somebody happened to regenerate that day.

    `strip` is where the branded centre strip SAT in this layout, as (y, height)
    (#970). It is recorded rather than left to be inferred because the editor
    otherwise reads it back out of the cells it is about to judge, and a row
    dragged down over the strip carries the inferred band down with it, so the
    check agrees with itself while the branding is being covered (L70). A layout
    with no strip, which is the reel's, records none.
    """
    document: dict[str, Any] = {"version": version, "cells": cells}
    if strip is not None:
        document["strip"] = {"y": strip[0], "h": strip[1]}
    Path(path).write_text(json.dumps(document), encoding="utf-8")


def read_layout_sidecar(path: Path | str) -> tuple[int | None, list[dict[str, Any]]]:
    """The (version, cells) in a sidecar, tolerating every shape on disk.

    A bare array is a sidecar written before the version existed, and its
    version is None rather than 0: not knowing which design made something is
    a different fact from knowing it was the first one.

    Never raises. A missing or corrupt sidecar means the crop editor falls back
    to the automatic layout, which is what it did before any of this.
    """
    try:
        doc = json.loads(Path(path).read_text(encoding="utf-8"))
    except (OSError, ValueError, TypeError):
        return None, []
    if isinstance(doc, list):
        return None, [c for c in doc if isinstance(c, dict)]
    if isinstance(doc, dict):
        cells = doc.get("cells")
        version = doc.get("version")
        return (
            version if isinstance(version, int) else None,
            [c for c in cells if isinstance(c, dict)] if isinstance(cells, list) else [],
        )
    return None, []
